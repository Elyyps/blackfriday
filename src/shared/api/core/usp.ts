import { IconLink } from "./icon-link";

export class USP {
  public uniqueSellingPoints: IconLink[];
}
