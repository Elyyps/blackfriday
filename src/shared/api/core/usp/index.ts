export * from "./endpoints";
export * from "./generate-dummy-data";
export * from "./usp";
