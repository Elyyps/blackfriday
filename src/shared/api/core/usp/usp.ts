import { IconLink } from "../icon-link";

export class UspModule {
  public uniqueSellingPoints: IconLink[];
}
