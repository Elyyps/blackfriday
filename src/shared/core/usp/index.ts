export * from "./usp.component";
export * from "./container/usp.container";
