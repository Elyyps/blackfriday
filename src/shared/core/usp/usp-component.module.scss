.usp {
  font-size: 13px;
  @media (max-width: $breakpoint-small-max) {
    font-size: 12px;
  }
  @media (max-width: 768px) {
    font-size: 13px;
  }
  font-weight: 600;
  display: flex;
  justify-content: center;
  width: 100%;
  justify-content: space-around;
  @media (max-width: 768px) {
    justify-content: center;
  }
  &__item {
    text-align: center;
  }
  &__link {
    color: $gray-light;
    text-decoration: none;
    display: inline-flex;
    align-items: center;
    text-align: left;
    &:hover {
      color: $gray;
      text-decoration: none;
      span {
        fill: $gray;
      }
    }
  }
  .icon-svg {
    margin-right: 8px;
    path {
      fill: currentColor;
    }
  }
  .usp__item {
    -webkit-transition: all 0.5s;
    transition: all 0.5s;
  }
  @media (max-width: 768px) {
    .usp__item {
      position: absolute;
      display: flex;
    }
    margin-bottom: 16px;
  }
  @media (max-width: 800px) {
    .usp__item {
      padding: 0 19px;
    }
  }
}
