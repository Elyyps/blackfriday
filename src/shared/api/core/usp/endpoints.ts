import { uniqueSellingPoints } from "./generate-dummy-data";
import { USP } from "./usp";

export const generateUSP = (): USP => uniqueSellingPoints;
