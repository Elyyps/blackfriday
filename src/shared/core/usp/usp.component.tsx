import React, { useState, useEffect } from "react";
import styles from "./usp-component.module.scss";
import { Link } from "react-router-dom";
import { IconComponent } from "@app/core";
import { USP as USP_CLASS } from "@app/api/core/usp/usp";

interface IUSPProps {
  uspModule: USP_CLASS;
}

let timerInterval: number;
const timerIntervalDuration = 3000;
const mobileMaxWidth = 361;

const setOpacityInArray = (array: any[], selectedElement: number) =>
  array.map((item, index) => {
    index === selectedElement ? (item.style.opacity = 1) : (item.style.opacity = 0);
  });

const startFadingSlider = () => {
  const selector = document.getElementById("uspHolder");
  if (typeof selector !== "undefined" && selector !== null) {
    const selectorArray = Array.prototype.slice.call(selector.children);
    const selectedElement = 0;
    setOpacityInArray(selectorArray, selectedElement);

    return intervalSlider(selectedElement, selectorArray);
  }

  return "";
};

const intervalSlider = (selectedElement: number, array: any[]) => {
  let thisItem = selectedElement;
  timerInterval = window.setInterval(() => {
    if (thisItem + 1 > array.length - 1) {
      thisItem = 0;
    } else {
      thisItem += 1;
    }
    setOpacityInArray(array, thisItem);
  }, timerIntervalDuration);
};

const stopFadingSlider = () => {
  const selector = document.getElementById("uspHolder");
  clearInterval(timerInterval);
  if (typeof selector !== "undefined" && selector !== null) {
    const selectorArray = Array.prototype.slice.call(selector.children);
    selectorArray.map(item => {
      item.style.opacity = 1;
      item.style.display = "flex";
    });
  }
};

const USP = (props: IUSPProps) => {
  const [mobileWidth, setMobileWidth] = useState(false);
  const updateWindowDimensions = () => {
    if (window.innerWidth < mobileMaxWidth) {
      if (!mobileWidth) {
        setMobileWidth(true);
      }
    } else if (window.innerWidth > mobileMaxWidth) {
      if (!mobileWidth) {
        setMobileWidth(false);
      }
    }
  };

  useEffect(() => {
    if (window.innerWidth < mobileMaxWidth) {
      setMobileWidth(true);
      updateWindowDimensions();
    }
  }, []);

  useEffect(() => {
    if (mobileWidth) {
      startFadingSlider();
    } else {
      stopFadingSlider();
    }
  }, [mobileWidth]);

  useEffect(() => {
    window.addEventListener("resize", updateWindowDimensions);
  });

  return (
    <div id="uspHolder" className={styles["usp"]}>
      {props.uspModule.uniqueSellingPoints.map((item, index) => (
        <div key={index} data-key={index} className={styles["usp__item"]}>
          <Link to={item.link.url} className={styles["usp__link"]}>
            <IconComponent icon={item.icon} size="14px" fillColor="#777777" />
            <span style={{ marginLeft: "10px" }}>{item.link.title}</span>
          </Link>
        </div>
      ))}
    </div>
  );
};

export { USP };
