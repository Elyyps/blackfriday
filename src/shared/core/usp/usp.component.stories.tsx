import * as React from "react";

import { storiesOf } from "@storybook/react";
import { withA11y } from "@storybook/addon-a11y";
import { withProvider } from "@app/util";

import { generateUSP } from "@app/api/core/usp/endpoints";
import { USPContainer } from "./container";

storiesOf("Unique Selling Points (USP)", module)
  .addDecorator(withA11y)
  .addDecorator(withProvider)
  .add("Basic implementation", () => <USPContainer uspModule={generateUSP()} />);
