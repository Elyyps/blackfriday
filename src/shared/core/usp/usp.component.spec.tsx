import React from "react";
import { shallow } from "enzyme";
import { USP } from "./usp.component";
import { generateUSP } from "@app/api/core/usp/endpoints";

describe("[StarsRating]", () => {
  it("should render component without crashing", () => {
    const renderedComponent = shallow(<USP uspModule={generateUSP()} />);
    expect(renderedComponent).toMatchSnapshot();
  });
});
