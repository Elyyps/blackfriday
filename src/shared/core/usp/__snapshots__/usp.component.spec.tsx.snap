// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[StarsRating] should render component without crashing 1`] = `
<div
  className="usp"
  id="uspHolder"
>
  <div
    className="usp__item"
    key="0"
    style={
      Object {
        "display": "flex",
        "opacity": 0,
      }
    }
  >
    <Link
      className="usp__link"
      to="./stores-overview"
    >
      <IconComponent
        fillColor="#777"
        icon="https://viper-development-images.s3-eu-west-1.amazonaws.com/blackFriday/footerIcons/shop.svg"
        size="14px"
      />
      <span
        style={
          Object {
            "marginLeft": "10px",
          }
        }
      >
        Overzicht van Ã¡lle black friday winkels
      </span>
    </Link>
  </div>
  <div
    className="usp__item"
    key="1"
    style={
      Object {
        "display": "flex",
        "opacity": 0,
      }
    }
  >
    <Link
      className="usp__link"
      to="./deals-overview"
    >
      <IconComponent
        fillColor="#777"
        icon="https://viper-development-images.s3-eu-west-1.amazonaws.com/blackFriday/footerIcons/clock.svg"
        size="14px"
      />
      <span
        style={
          Object {
            "marginLeft": "10px",
          }
        }
      >
        Altijd up-to-date van de laatste deals
      </span>
    </Link>
  </div>
  <div
    className="usp__item"
    key="2"
    style={
      Object {
        "display": "flex",
        "opacity": 0,
      }
    }
  >
    <Link
      className="usp__link"
      to="./"
    >
      <IconComponent
        fillColor="#777"
        icon="https://viper-development-images.s3-eu-west-1.amazonaws.com/blackFriday/footerIcons/megaphone.svg"
        size="14px"
      />
      <span
        style={
          Object {
            "marginLeft": "10px",
          }
        }
      >
        Mis niets van Black Friday 2019
      </span>
    </Link>
  </div>
</div>
`;
